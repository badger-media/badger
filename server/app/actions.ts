"use server";

import { checkSession, deleteSession } from "@/lib/auth";
import { db } from "@/lib/db";
import invariant from "@/lib/invariant";
import { redirect } from "next/navigation";

export async function DEV_promoteSelf() {
  invariant(process.env.NODE_ENV === "development", "Dev-only [1]");
  invariant(process.env.ENVIRONMENT !== "prod", "Dev-only [2]");
  const me = await checkSession();
  invariant(me, "No user");
  await db.user.update({
    where: {
      id: me.id,
    },
    data: {
      permissions: [...me.permissions, "SUDO"],
    },
  });
  await deleteSession();
  redirect("/");
}
