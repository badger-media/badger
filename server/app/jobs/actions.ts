"use server";

import { requirePermission } from "@/lib/auth";
import { db } from "@/lib/db";
import { dispatchJobForJobrunner } from "@/lib/jobs";
import { revalidatePath } from "next/cache";

export async function doResetJob(baseJobID: number) {
  await requirePermission("ManageJobs");
  await db.baseJob.update({
    where: {
      id: baseJobID,
    },
    data: {
      startedAt: null,
      completedAt: null,
      workerId: null,
      externalJobProvider: null,
      externalJobID: null,
      state: "Pending",
    },
  });
  await dispatchJobForJobrunner(baseJobID);
  revalidatePath("/jobs");
  return { ok: true };
}
