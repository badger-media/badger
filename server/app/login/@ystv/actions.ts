"use server";

import { SignInSchema } from "@/app/login/@ystv/schema";
import { z } from "zod";
import { FormResponse } from "@/components/Form";
import { redirect } from "next/navigation";
import { InvalidCredentials } from "@/lib/auth/types";
import { DummyTestAuth } from "@/lib/auth/dummyTest";
import { YSTVAuth } from "@/lib/auth/ystv";
import invariant from "@/lib/invariant";
import { SignInResult, doSignIn } from "@/lib/auth";
import { useDummyTestAuth } from "@badger/feature-flags";

function determineProvider() {
  if (useDummyTestAuth) {
    invariant(
      process.env.NODE_ENV !== "production" || process.env.E2E_TEST === "true",
      "Cannot enable dummy test auth in production",
    );
    console.warn("Using dummy test auth - do *not* use this in production!");
    return DummyTestAuth;
  }
  if (process.env.YSTV_SSO_USERNAME && process.env.YSTV_SSO_PASSWORD) {
    return YSTVAuth;
  }
  invariant(
    false,
    "No authentication configured. Set USE_DUMMY_TEST_AUTH=true or set YSTV_SSO_USERNAME and YSTV_SSO_PASSWORD",
  );
}

const provider = determineProvider();

export async function handleSignIn(
  data: z.infer<typeof SignInSchema>,
): Promise<FormResponse> {
  try {
    const creds = await provider.checkCredentials(data.username, data.password);
    const loginResult = await doSignIn(provider.id, creds);
    switch (loginResult) {
      case SignInResult.Success:
        break;
      case SignInResult.CreatedInactive:
      case SignInResult.Inactive:
        return {
          ok: false,
          errors: {
            root: "Your account is not yet active. Please contact the Computing Team.",
          },
        };
    }
    if (data.return) {
      if (data.return.startsWith("/")) {
        const url = new URL(data.return, process.env.PUBLIC_URL);
        redirect(url.toString());
      }
      redirect(data.return);
    }
    return { ok: true };
  } catch (e) {
    if (e instanceof InvalidCredentials) {
      return {
        ok: false,
        errors: {
          root: "Incorrect username or password",
        },
      };
    }
    throw e;
  }
}
