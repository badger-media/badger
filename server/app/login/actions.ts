"use server";

import { SignInSchema } from "@/app/login/schema";
import { z } from "zod";
import { FormResponse } from "@/components/Form";
import { redirect } from "next/navigation";
import { doSignIn } from "@/lib/auth";
import { InvalidCredentials } from "@/lib/auth/types";

export async function handleSignIn(
  data: z.infer<typeof SignInSchema>,
): Promise<FormResponse> {
  try {
    await doSignIn(data.username, data.password);
    if (data.return) {
      if (data.return.startsWith("/")) {
        const url = new URL(data.return, process.env.PUBLIC_URL);
        redirect(url.toString());
      }
      redirect(data.return);
    }
    return { ok: true };
  } catch (e) {
    if (e instanceof InvalidCredentials) {
      return {
        ok: false,
        errors: {
          root: "Incorrect username or password",
        },
      };
    }
    throw e;
  }
}
