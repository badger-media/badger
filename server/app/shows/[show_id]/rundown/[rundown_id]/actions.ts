"use server";

import { db } from "@/lib/db";
import { revalidatePath } from "next/cache";
import { notFound } from "next/navigation";

export async function revalidateIfChanged(rundownID: number, version: number) {
  const rundown = await db.rundown.findUnique({
    where: { id: rundownID },
    select: {
      show: {
        select: { version: true, id: true },
      },
    },
  });
  if (!rundown) {
    notFound();
  }
  if (rundown.show.version !== version) {
    revalidatePath(`/shows/${rundown.show.id}/rundown/${rundownID}`);
  }
}
