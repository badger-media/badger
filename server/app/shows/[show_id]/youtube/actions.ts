"use server";

import { db } from "@/lib/db";
import { youtube_v3 } from "googleapis";
import { z } from "zod";
import { createStreamsPayloadSchema } from "./schema";
import { FormResponse } from "@/components/Form";
import { checkSession } from "@/lib/auth";
import { cookies } from "next/headers";
import { OAuth2Client } from "google-auth-library";
import { ConnectionTarget } from "@bowser/prisma/client";
import { redirect } from "next/navigation";
import invariant from "@/lib/invariant";
import {
  getConnectionAccessToken,
  makeGoogleOauthClient,
} from "@/lib/connections";

async function getAccessTokenForCurrentUser() {
  // If we already have an access token in cookies, use that
  const cookie = await cookies().get("google_access_token");
  if (cookie?.value) {
    return cookie.value;
  }
  // Now check if we already have a refresh token, and if so, use that to get an access token.
  const user = await checkSession();
  const connection = await db.connection.findFirst({
    where: {
      target: ConnectionTarget.google,
      userId: user!.id,
    },
  });
  if (!connection) {
    // Damn and blast.
    redirect("/connect/google");
  }

  const client = new OAuth2Client({
    clientId: process.env.GOOGLE_CLIENT_ID!,
    clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    redirectUri: process.env.GOOGLE_CONNECT_REDIRECT_URI!,
    credentials: {
      refresh_token: connection.refreshToken,
    },
  });
  const accessToken = await client.getAccessToken();
  invariant(accessToken.token, "failed to refresh token");
  await cookies().set("google_access_token", accessToken.token, {
    secure: process.env.NODE_ENV === "production",
    httpOnly: true,
    expires: new Date(Date.now() + 60 * 60 * 1000),
    sameSite: "strict",
  });
  return accessToken.token;
}

export async function doCreateStreams(
  dataRaw: z.infer<typeof createStreamsPayloadSchema>,
): Promise<FormResponse> {
  const data = createStreamsPayloadSchema.parse(dataRaw);
  const me = await checkSession();
  invariant(me, "no current user");

  await db.$transaction(async ($db) => {
    // Lock the row for the duration of this transaction
    await $db.$queryRaw`SELECT id FROM shows WHERE id = ${data.show_id} FOR UPDATE`;
    const show = await $db.show.findUniqueOrThrow({
      where: {
        id: data.show_id,
      },
      include: {
        rundowns: true,
        continuityItems: true,
      },
    });

    const token = await getConnectionAccessToken(
      ConnectionTarget.google,
      me.id,
    );
    if (!token) {
      redirect("/connect/google");
    }
    const client = makeGoogleOauthClient();
    client.setCredentials({
      access_token: token,
    });

    const yt = new youtube_v3.Youtube({
      auth: client,
    });
    console.log(yt);

    // Create the stream
    let streamID;
    if (show.ytStreamID) {
      console.log("Using existing stream", show.ytStreamID);
      streamID = show.ytStreamID;
    } else {
      const stream = await yt.liveStreams.insert({
        part: ["id", "snippet", "cdn"],
        requestBody: {
          snippet: {
            title: show.name,
          },
          cdn: {
            ingestionType: data.ingestionType,
            resolution: data.resolution,
            frameRate: data.frameRate,
          },
        },
      });
      await $db.show.update({
        where: {
          id: show.id,
        },
        data: {
          ytStreamID: stream.data.id!,
        },
      });
      streamID = stream.data.id!;
    }
    // Now create the broadcasts
    for (const item of data.items) {
      if (!item.enabled) {
        continue;
      }
      if (item.rundownID) {
        const rd = show.rundowns.find((x) => x.id === item.rundownID);
        if (rd?.ytBroadcastID?.length) {
          console.log("Skipping broadcast creation for rundown", rd.id);
          continue;
        }
      } else if (item.continuityItemID) {
        const ci = show.continuityItems.find(
          (x) => x.id === item.continuityItemID,
        );
        if (ci?.ytBroadcastID?.length) {
          console.log("Skipping broadcast creation for continuity item", ci.id);
          continue;
        }
      } else if (item.isShowBroadcast) {
        if (show.ytBroadcastID?.length) {
          console.log("Skipping broadcast creation for show", show.id);
          continue;
        }
      }

      const broadcast = await yt.liveBroadcasts.insert({
        part: ["id", "snippet", "status", "contentDetails"],
        requestBody: {
          snippet: {
            title: item.title,
            description: item.description,
            scheduledStartTime: item.start.toISOString(),
            scheduledEndTime: item.end.toISOString(),
          },
          status: {
            privacyStatus: item.visibility,
            selfDeclaredMadeForKids: false,
          },
          contentDetails: {
            enableAutoStart: false,
            enableAutoStop: false,
            enableClosedCaptions: true,
            enableDvr: true,
            enableEmbed: data.enableEmbed,
            recordFromStart: true,
          },
        },
      });
      // TODO[BOW-132]: Set thumbnail
      // await yt.thumbnails.set({
      //   videoId: broadcast.data.id!,
      //   media: {
      //     mimeType: "image/jpeg",
      //     body: item.thumbnail,
      //   }
      // });
      await yt.liveBroadcasts.bind({
        id: broadcast.data.id!,
        part: ["id"],
        streamId: streamID,
      });
      if (item.rundownID) {
        await $db.rundown.update({
          where: {
            id: item.rundownID,
          },
          data: {
            ytBroadcastID: broadcast.data.id!,
          },
        });
      } else if (item.continuityItemID) {
        await $db.continuityItem.update({
          where: {
            id: item.continuityItemID,
          },
          data: {
            ytBroadcastID: broadcast.data.id!,
          },
        });
      } else if (item.isShowBroadcast) {
        await $db.show.update({
          where: {
            id: show.id,
          },
          data: {
            ytBroadcastID: broadcast.data.id!,
          },
        });
      }
    }
  });

  return { ok: true };
}
