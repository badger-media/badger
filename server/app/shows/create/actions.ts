"use server";

import { schema } from "./schema";
import { revalidatePath } from "next/cache";
import { db } from "@/lib/db";
import { FormResponse } from "@/components/Form";
import { zodErrorResponse } from "@/components/FormServerHelpers";
import { z } from "zod";

export async function create(
  data: z.infer<typeof schema>,
): Promise<FormResponse> {
  "use server";
  const result = schema.safeParse(data);
  if (!result.success) {
    return zodErrorResponse(result.error);
  }
  const res = await db.show.create({
    data: {
      name: result.data.name,
      start: result.data.start,
    },
  });
  revalidatePath(`/`);
  revalidatePath(`/shows/${res.id}`);
  return { ok: true, id: res.id }; // TODO: this should not be necessary, a redirect() should suffice
}
