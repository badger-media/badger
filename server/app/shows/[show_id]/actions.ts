"use server";

import { db } from "@/lib/db";
import { revalidatePath } from "next/cache";
import { notFound } from "next/navigation";
import { z } from "zod";
import { editContinuityItemSchema } from "./schema";
import { FormResponse } from "@/components/Form";
import { zodErrorResponse } from "@/components/FormServerHelpers";
import {
  MediaFileSourceType,
  MetadataTargetType,
  Prisma,
} from "@bowser/prisma/client";
import { escapeRegExp } from "lodash";

import { dispatchJobForJobrunner } from "@/lib/jobs";

export async function addItem(
  showID: number,
  type: "rundown" | "continuity_item",
  name: string,
  duration?: number,
) {
  await db.$transaction(async ($db) => {
    const highestOrder = await $db.$queryRaw<[{ order: number }]>`
      SELECT COALESCE(
        GREATEST(
          (SELECT MAX("order") FROM "rundowns" WHERE "showId" = ${showID}),
          (SELECT MAX("order") FROM "continuity_items" WHERE "showId" = ${showID})
        ),
        -1 -- to ensure that the first item is 0-indexed
      ) AS "order"
    `;
    const order = highestOrder[0].order + 1;
    if (type === "rundown") {
      await $db.rundown.create({
        data: {
          name,
          order,
          show: {
            connect: { id: showID },
          },
        },
      });
    } else {
      await $db.continuityItem.create({
        data: {
          name,
          order,
          durationSeconds: duration ?? 0,
          show: {
            connect: { id: showID },
          },
        },
      });
    }
    await $db.show.update({
      where: {
        id: showID,
      },
      data: {
        version: {
          increment: 1,
        },
      },
    });
  });
  revalidatePath(`/shows/${showID}`);
  return { ok: true };
}

// rundowns are handled on their own page (or at least will be - TODO)
export async function editContinuityItem(
  raw: z.infer<typeof editContinuityItemSchema>,
): Promise<FormResponse> {
  const data = editContinuityItemSchema.safeParse(raw);
  if (!data.success) {
    return zodErrorResponse(data.error);
  }
  const res = await db.continuityItem.update({
    where: {
      id: data.data.itemID,
    },
    data: {
      name: data.data.name,
      durationSeconds: data.data.duration,
      show: {
        update: {
          version: {
            increment: 1,
          },
        },
      },
    },
  });
  revalidatePath(`/shows/${res.showId}`);
  return { ok: true };
}

export async function deleteItem(
  showID: number,
  itemType: "rundown" | "continuity_item",
  itemId: number,
) {
  await db.$transaction(async ($db) => {
    let oldItem;
    if (itemType === "rundown") {
      oldItem = await $db.rundown.delete({
        where: {
          id: itemId,
        },
      });
    } else if (itemType === "continuity_item") {
      oldItem = await $db.continuityItem.delete({
        where: {
          id: itemId,
        },
      });
    } else {
      throw new Error("Invalid item type");
    }
    // Ensure order remains contiguous
    await $db.rundown.updateMany({
      where: {
        showId: showID,
        order: {
          gt: oldItem.order,
        },
      },
      data: {
        order: {
          decrement: 1,
        },
      },
    });
    await $db.continuityItem.updateMany({
      where: {
        showId: showID,
        order: {
          gt: oldItem.order,
        },
      },
      data: {
        order: {
          decrement: 1,
        },
      },
    });
    await $db.show.update({
      where: {
        id: showID,
      },
      data: {
        version: {
          increment: 1,
        },
      },
    });
    await ensureContiguousDEV(showID, $db);
  });
  revalidatePath(`/shows/${showID}`);
}

export async function reorderShowItems(
  showID: number,
  itemType: "rundown" | "continuity_item",
  itemId: number,
  newOrder: number,
) {
  await db.$transaction(async ($db) => {
    // TODO: the rundowns and continuity_items tables don't have unique constraints.
    // See the comment in schema.prisma for the background on this.

    const show = await $db.show.findUnique({
      where: {
        id: showID,
      },
      include: {
        rundowns: true,
        continuityItems: true,
      },
    });
    if (!show) {
      notFound();
    }

    // If the item is moving down, we need to move all items between the old and new position up one.
    // If the item is moving up, we need to move all items between the new and old position down one.
    // Remember that order is ascending, so moving an item "down" actually means incrementing its position.
    const oldItem =
      itemType === "rundown"
        ? show.rundowns.find((x) => x.id === itemId)
        : show.continuityItems.find((x) => x.id === itemId);
    if (!oldItem) {
      throw new Error("Invalid item ID");
    }
    const oldOrder = oldItem.order;
    console.debug(
      `Moving ${itemType} ${itemId} from ${oldOrder} to ${newOrder}`,
    );
    if (oldOrder < newOrder) {
      // Moving down, so the item's order will be incremented
      await $db.rundown.updateMany({
        where: {
          showId: showID,
          order: {
            gte: oldOrder,
            lte: newOrder,
          },
        },
        data: {
          order: {
            decrement: 1,
          },
        },
      });
      await $db.continuityItem.updateMany({
        where: {
          showId: showID,
          order: {
            gte: oldOrder,
            lte: newOrder,
          },
        },
        data: {
          order: {
            decrement: 1,
          },
        },
      });
    } else {
      // Moving up, so the item's order will be decremented
      await $db.rundown.updateMany({
        where: {
          showId: showID,
          order: {
            lte: oldOrder,
            gte: newOrder,
          },
        },
        data: {
          order: {
            increment: 1,
          },
        },
      });
      await $db.continuityItem.updateMany({
        where: {
          showId: showID,
          order: {
            lte: oldOrder,
            gte: newOrder,
          },
        },
        data: {
          order: {
            increment: 1,
          },
        },
      });
    }

    if (itemType === "rundown") {
      await $db.rundown.update({
        where: {
          id: itemId,
        },
        data: {
          order: newOrder,
        },
      });
    } else if (itemType === "continuity_item") {
      await $db.continuityItem.update({
        where: {
          id: itemId,
        },
        data: {
          order: newOrder,
        },
      });
    } else {
      throw new Error("Invalid item type");
    }

    await $db.show.update({
      where: {
        id: showID,
      },
      data: {
        version: {
          increment: 1,
        },
      },
    });

    await ensureContiguousDEV(showID, $db);
  });
  revalidatePath(`/shows/${showID}`);
  return { ok: true };
}

export async function setMetaValue(
  showID: number,
  metaID: number,
  newValue: Prisma.InputJsonValue,
): Promise<FormResponse> {
  await db.metadata.update({
    where: {
      field: {
        target: MetadataTargetType.Show,
      },
      showId: showID,
      id: metaID,
    },
    data: {
      value: newValue,
    },
  });
  revalidatePath(`/shows/${showID}`);
  return { ok: true };
}

export async function addMeta(
  showID: number,
  fieldID: number,
  newValue: Prisma.InputJsonValue,
): Promise<FormResponse> {
  await db.metadata.create({
    data: {
      fieldId: fieldID,
      showId: showID,
      value: newValue,
    },
  });
  revalidatePath(`/shows/${showID}`);
  return { ok: true };
}

async function ensureContiguousDEV(
  showID: number,
  $db: Parameters<Parameters<typeof db.$transaction>[0]>[0],
) {
  if (process.env.NODE_ENV !== "production") {
    // Sanity check: ensure all items have a contiguous order, otherwise this will break next time
    const show = await $db.show.findUnique({
      where: {
        id: showID,
      },
      include: {
        rundowns: true,
        continuityItems: true,
      },
    });
    if (!show) {
      notFound();
    }
    const allItems = [...show.rundowns, ...show.continuityItems]
      .map((x) => x.order)
      .sort((a, b) => a - b);
    if (allItems.length >= 2) {
      for (let i = 1; i < allItems.length - 1; i++) {
        if (allItems[i] !== allItems[i - 1] + 1) {
          throw new Error(
            `Invariant violation: non-contiguous order for items of show ${showID}; item ${i} has order ${
              allItems[i]
            } but item ${i - 1} has order ${allItems[i - 1]}`,
          );
        }
      }
    }
    if (
      allItems.length > 0 &&
      allItems[allItems.length - 1] !== allItems.length - 1
    ) {
      throw new Error(
        `Invariant violation: non-contiguous order for items of show ${showID}; last item has order ${
          allItems[allItems.length - 1]
        } but there are ${allItems.length} items`,
      );
    }
  }
}

export async function processUploadForContinuityItem(
  itemID: number,
  fileName: string,
  uploadURL: string,
) {
  const item = await db.continuityItem.findUnique({
    where: {
      id: itemID,
    },
  });
  if (!item) {
    throw new Error("Invalid item ID");
  }

  // Sanity check to ensure it was really uploaded where we expected
  if (!uploadURL.startsWith(process.env.TUS_ENDPOINT!)) {
    throw new Error("Invalid upload URL");
  }

  const baseJobID = await db.$transaction(async ($db) => {
    await $db.media.deleteMany({
      where: {
        continuityItem: {
          id: itemID,
        },
      },
    });
    const res = await $db.media.create({
      data: {
        name: fileName,
        durationSeconds: 0,
        rawPath: "",
        continuityItem: {
          connect: {
            id: itemID,
          },
        },
        process_jobs: {
          create: {
            sourceType: MediaFileSourceType.Tus,
            source: uploadURL.replace(
              // Strip off the Tus endpoint prefix so the source is just the ID
              new RegExp(`^${escapeRegExp(process.env.TUS_ENDPOINT!)}/?`),
              "",
            ),
            base_job: {
              create: {},
            },
          },
        },
      },
      include: {
        process_jobs: true,
      },
    });
    await $db.show.update({
      where: {
        id: item.showId,
      },
      data: {
        version: {
          increment: 1,
        },
      },
    });
    return res.process_jobs[0].base_job_id;
  });
  await dispatchJobForJobrunner(baseJobID);
  revalidatePath(`/shows/${item.showId}`);
  return { ok: true };
}
