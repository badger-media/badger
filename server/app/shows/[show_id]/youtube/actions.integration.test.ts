import { db } from "@/lib/db";
import { test, vi, beforeEach, expect, it } from "vitest";
import { doCreateStreams } from "./actions";
import { integrate } from "@bowser/testing";
import { youtube_v3 } from "googleapis";
import { Show } from "@bowser/prisma/client";

vi.mock("server-only", () => ({}));
vi.mock("next/cache", () => ({ revalidatePath: () => {} }));
vi.mock("@/lib/auth", () => ({
  checkSession: () => Promise.resolve({}),
}));
vi.mock("@/lib/connections", async () => {
  const actual =
    await vi.importActual<typeof import("@/lib/connections")>(
      "@/lib/connections",
    );
  return {
    ...actual,
    getConnectionAccessToken: () => "GOOGLE_ACCESS_TOKEN",
  };
});
vi.mock("googleapis", async () => {
  const Youtube = vi.fn();

  let streamID = 0;
  let broadcastID = 0;

  Youtube.prototype.liveStreams = {
    insert: vi.fn(() => ({
      data: {
        id: "test-stream-" + streamID++,
      },
    })),
  };
  Youtube.prototype.liveBroadcasts = {
    insert: vi.fn(() => ({
      data: {
        id: "test-broadcast-" + broadcastID++,
      },
    })),
    bind: vi.fn(),
  };

  return { youtube_v3: { Youtube } };
});

const TEST_TIME = new Date("2023-07-21T16:46:35.036Z");

integrate("youtube/doCreateStreams", async () => {
  let testShow: Show;
  beforeEach(async () => {
    await Promise.all(
      ["shows", "metadata"].map((table) =>
        db.$executeRawUnsafe(`DELETE FROM ${table}`),
      ),
    );
    testShow = await db.show.create({
      data: {
        name: "Test Show",
        start: TEST_TIME,
        rundowns: {
          createMany: {
            data: [
              {
                name: "Test 1",
                order: 0,
              },
            ],
          },
        },
        continuityItems: {
          createMany: {
            data: [
              {
                name: "Test 2",
                durationSeconds: 0,
                order: 1,
              },
            ],
          },
        },
      },
    });
  });

  it("works", async () => {
    await doCreateStreams({
      show_id: testShow.id,
      resolution: "1080p",
      frameRate: "30fps",
      ingestionType: "rtmp",
      items: [
        {
          title: "Test Show",
          description: "",
          start: TEST_TIME,
          end: TEST_TIME,
          enabled: true,
          visibility: "public",
          isShowBroadcast: true,
        },
        {
          title: "Test 1",
          description: "",
          start: TEST_TIME,
          end: TEST_TIME,
          enabled: true,
          visibility: "public",
          rundownID: 1,
        },
        {
          title: "Test 2",
          description: "",
          start: TEST_TIME,
          end: TEST_TIME,
          enabled: false,
          visibility: "public",
          continuityItemID: 1,
        },
      ],
    });
    const yt = vi.mocked(youtube_v3.Youtube).mock.instances[0];
    expect(yt.liveStreams.insert).toHaveBeenCalled();
    expect(
      vi.mocked(yt.liveStreams.insert).mock.lastCall![0],
    ).toMatchSnapshot();

    expect(yt.liveBroadcasts.insert).toHaveBeenCalledTimes(2);
    expect(
      vi.mocked(yt.liveBroadcasts.insert).mock.calls[0][0],
    ).toMatchSnapshot();
    expect(
      vi.mocked(yt.liveBroadcasts.bind).mock.calls[0][0],
    ).toMatchSnapshot();
    expect(
      vi.mocked(yt.liveBroadcasts.insert).mock.calls[1][0],
    ).toMatchSnapshot();
    expect(
      vi.mocked(yt.liveBroadcasts.bind).mock.calls[1][0],
    ).toMatchSnapshot();

    const dbShow = await db.show.findFirstOrThrow({
      where: { id: 1 },
      include: { rundowns: true, continuityItems: true },
    });
    expect(dbShow.ytStreamID).toEqual("test-stream-0");
    expect(dbShow.ytBroadcastID).toEqual("test-broadcast-0");
    expect(dbShow.rundowns[0].ytBroadcastID).toEqual("test-broadcast-1");
    expect(dbShow.continuityItems[0].ytBroadcastID).toBeNull();
  });
});
