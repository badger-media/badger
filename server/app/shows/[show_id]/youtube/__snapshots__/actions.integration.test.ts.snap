// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`youtube/doCreateStreams works 1`] = `
{
  "part": [
    "id",
    "snippet",
    "cdn",
  ],
  "requestBody": {
    "cdn": {
      "frameRate": "30fps",
      "ingestionType": "rtmp",
      "resolution": "1080p",
    },
    "snippet": {
      "title": "Test Show",
    },
  },
}
`;

exports[`youtube/doCreateStreams works 2`] = `
{
  "part": [
    "id",
    "snippet",
    "status",
    "contentDetails",
  ],
  "requestBody": {
    "contentDetails": {
      "enableAutoStart": false,
      "enableAutoStop": false,
      "enableClosedCaptions": true,
      "enableDvr": true,
      "enableEmbed": true,
      "recordFromStart": true,
    },
    "snippet": {
      "description": "",
      "scheduledEndTime": "2023-07-21T16:46:35.036Z",
      "scheduledStartTime": "2023-07-21T16:46:35.036Z",
      "title": "Test Show",
    },
    "status": {
      "privacyStatus": "public",
      "selfDeclaredMadeForKids": false,
    },
  },
}
`;

exports[`youtube/doCreateStreams works 3`] = `
{
  "id": "test-broadcast-0",
  "part": [
    "id",
  ],
  "streamId": "test-stream-0",
}
`;

exports[`youtube/doCreateStreams works 4`] = `
{
  "part": [
    "id",
    "snippet",
    "status",
    "contentDetails",
  ],
  "requestBody": {
    "contentDetails": {
      "enableAutoStart": false,
      "enableAutoStop": false,
      "enableClosedCaptions": true,
      "enableDvr": true,
      "enableEmbed": true,
      "recordFromStart": true,
    },
    "snippet": {
      "description": "",
      "scheduledEndTime": "2023-07-21T16:46:35.036Z",
      "scheduledStartTime": "2023-07-21T16:46:35.036Z",
      "title": "Test 1",
    },
    "status": {
      "privacyStatus": "public",
      "selfDeclaredMadeForKids": false,
    },
  },
}
`;

exports[`youtube/doCreateStreams works 5`] = `
{
  "id": "test-broadcast-1",
  "part": [
    "id",
  ],
  "streamId": "test-stream-0",
}
`;
